# pandas 也是個爬蟲好工具呢!

In [1]:
import pandas as pd
import time

# 自定義方法

In [9]:
def get_station_data(YEAR, MONTH, DAY):
    time.sleep(0.1)
    base_url = f"http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR={YEAR}&MONTH={MONTH}&FROM={DAY}00&TO={DAY}00&STNM=58968"
    print(base_url)
    raw_data = pd.read_csv(base_url)
    try:
        data = raw_data["<HTML>"]
        return data, base_url, True
    except:
        print(f"[SYSTEM] {YEAR}{MONTH}{DAY} No Record")
        print(base_url)
        data = None
        return data, base_url, False

def detect_start_line(line, i):
    x = line.split()
    if x[0] == "1000.0": 
        return True, i
    else:
        return False, 0

def detect_final_line(line, i):
    x = line.split()
    if x[0] == "300.0":
        return True, i+1
    else:
        return False, 0

def test_complete(line):
    x = line.split()
    if len(x) == 11: return True
    
def test_dataframe_and_save(dataframe, YEAR, MONTH, DAY):
    if dataframe_data.shape == (7,11):
        dataframe_data.to_csv(f"./data/{YEAR}-{MONTH}-{DAY}.csv")
    else:
        print(f"[SYSTEM] {YEAR}{MONTH}{DAY} Incomplete Data")                

def list_to_dataframe(list):
    return pd.DataFrame(list)

In [6]:
# 決定資料區間
YEARS = [2015,2016,2017,2018]
MONTHS= ["05","06","07","08","09","10"]
DAY_30= ["01","02","03","04","05","06","07","08","09","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30"]
DAY_31= ["01","02","03","04","05","06","07","08","09","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31"]

# 下面的 block 會將資料 append 到一個 sounding_data_list 裡面

In [10]:
YEAR = 2018

for MONTH in MONTHS:
    if MONTH in ["05","07","08","10"]: # 大月
        for DAY in DAY_31:
            flag = True
            data, url, flag = get_station_data(YEAR,MONTH,DAY)
            if flag == False: continue # 可以回頭看看 def get_station_data
        
            # 找到資料的第一行: i.e. hPa為 1000.0
            for i in range(len(data)):
                flag = False
                line = data[i]
                flag, START_LINE = detect_start_line(line, i)
                if flag != False: break

            # 找到資料的最後一行: i.e. hPa為 300.0
            for i in range(len(data)):
                flag = False
                line = data[i]
                flag, FINAL_LINE = detect_final_line(line, i)
                if flag != False: break
            
            # 用來儲存資料的容器，一天會有一筆，所以每一天都要清空一次   
            sounding_data_list = []
        
            for i in range(START_LINE, FINAL_LINE):
                line = data[i]
                if test_complete(line): sounding_data_list.append(line.split())
            
            # 將資料轉成　dataframe 的形式
            dataframe_data = list_to_dataframe(sounding_data_list)
            test_dataframe_and_save(dataframe_data, YEAR, MONTH, DAY)
            
    else: # 小月(6月和9月)，下面跟上面大同小異，通常會建議模組化，比較乾淨，不要學我XD
        for DAY in DAY_30:
            flag == True
            data, url, flag = get_station_data(YEAR,MONTH,DAY)
            if flag == False: continue
        
            # 找到資料的第一行: i.e. hPa為 1000.0
            for i in range(len(data)):
                flag = False
                line = data[i]
                flag, START_LINE = detect_start_line(line, i)
                if flag != False: break

            # 找到資料的最後一行: i.e. hPa為 300.0
            for i in range(len(data)):
                flag = False
                line = data[i]
                flag, FINAL_LINE = detect_final_line(line, i)
                if flag != False: break

            # 用來儲存資料的容器，一天會有一筆，所以每一天都要清空一次   
            sounding_data_list = []
        
            for i in range(START_LINE, FINAL_LINE):
                line = data[i]
                if test_complete(line): sounding_data_list.append(line.split())

            # 將資料轉成　dataframe 的形式
            dataframe_data = list_to_dataframe(sounding_data_list)
            test_dataframe_and_save(dataframe_data, YEAR, MONTH, DAY)

http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR=2018&MONTH=05&FROM=0100&TO=0100&STNM=58968
http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR=2018&MONTH=05&FROM=0200&TO=0200&STNM=58968
http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR=2018&MONTH=05&FROM=0300&TO=0300&STNM=58968
http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR=2018&MONTH=05&FROM=0400&TO=0400&STNM=58968
http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR=2018&MONTH=05&FROM=0500&TO=0500&STNM=58968
http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR=2018&MONTH=05&FROM=0600&TO=0600&STNM=58968
http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR=2018&MONTH=05&FROM=0700&TO=0700&STNM=58968
http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR=2018&MONTH=05&FROM=0800&TO=0800&STNM=58968
http://weather.uwyo.edu/cgi-bin/sounding

http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR=2018&MONTH=07&FROM=0800&TO=0800&STNM=58968
[SYSTEM] 20180708 No Record
http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR=2018&MONTH=07&FROM=0800&TO=0800&STNM=58968
http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR=2018&MONTH=07&FROM=0900&TO=0900&STNM=58968
http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR=2018&MONTH=07&FROM=1000&TO=1000&STNM=58968
http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR=2018&MONTH=07&FROM=1100&TO=1100&STNM=58968
http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR=2018&MONTH=07&FROM=1200&TO=1200&STNM=58968
http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR=2018&MONTH=07&FROM=1300&TO=1300&STNM=58968
http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR=2018&MONTH=07&FROM=1400&TO=1400&STNM=58968
http://weath

http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR=2018&MONTH=09&FROM=1300&TO=1300&STNM=58968
http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR=2018&MONTH=09&FROM=1400&TO=1400&STNM=58968
http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR=2018&MONTH=09&FROM=1500&TO=1500&STNM=58968
http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR=2018&MONTH=09&FROM=1600&TO=1600&STNM=58968
http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR=2018&MONTH=09&FROM=1700&TO=1700&STNM=58968
http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR=2018&MONTH=09&FROM=1800&TO=1800&STNM=58968
http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR=2018&MONTH=09&FROM=1900&TO=1900&STNM=58968
http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR=2018&MONTH=09&FROM=2000&TO=2000&STNM=58968
http://weather.uwyo.edu/cgi-bin/sounding

# 2018 小結報[PASS]
* [SYSTEM] 20180511 No Record
* [SYSTEM] 20180708 No Record
* [SYSTEM] 20180813 Incomplete Data
* [SYSTEM] 20181017 Incomplete Data
* [SYSTEM] 20181020 No Record

## 5月至10月共有 184 天
### 記錄到問題資料有 5 筆
* incomplet data: 2
* no record: 3
### 記錄到正常資料有 179 筆

# 2017 小結報[PASS]
* [SYSTEM] 20170501 No Record
* [SYSTEM] 20170502 No Record
* [SYSTEM] 20170503 No Record
* [SYSTEM] 20170504 No Record
* [SYSTEM] 20170505 No Record
* [SYSTEM] 20170506 No Record
* [SYSTEM] 20170507 No Record
* [SYSTEM] 20170508 No Record
* [SYSTEM] 20170509 No Record
* [SYSTEM] 20170510 No Record
* [SYSTEM] 20170511 No Record
* [SYSTEM] 20170512 No Record
* [SYSTEM] 20170513 No Record
* [SYSTEM] 20170514 No Record
* [SYSTEM] 20170515 No Record
* [SYSTEM] 20170516 No Record
* [SYSTEM] 20170517 No Record
* [SYSTEM] 20170518 No Record
* [SYSTEM] 20170519 No Record
* [SYSTEM] 20170520 No Record
* [SYSTEM] 20170521 No Record
* [SYSTEM] 20170522 No Record
* [SYSTEM] 20170523 No Record
* [SYSTEM] 20170524 No Record
* [SYSTEM] 20170525 No Record
* [SYSTEM] 20170526 No Record
* [SYSTEM] 20170527 No Record
* [SYSTEM] 20170528 No Record
* [SYSTEM] 20170529 No Record
* [SYSTEM] 20170530 No Record
* [SYSTEM] 20170531 No Record
* [SYSTEM] 20170602 Incomplete Data
* [SYSTEM] 20170730 Incomplete Data
* [SYSTEM] 20171003 No Record

## 5月至10月共有 184 天
### 記錄到問題資料有 34 筆
* incomplet data: 2
* no record: 32
### 記錄到正常資料有 150 筆

# 2016 小結報[PASS]
* [SYSTEM] 20160501 No Record
* [SYSTEM] 20160502 No Record
* [SYSTEM] 20160503 No Record
* [SYSTEM] 20160527 No Record
* [SYSTEM] 20160529 No Record
* [SYSTEM] 20160606 No Record
* [SYSTEM] 20160609 No Record
* [SYSTEM] 20160710 No Record
* [SYSTEM] 20160711 No Record
* [SYSTEM] 20160727 No Record
* [SYSTEM] 20160728 No Record
* [SYSTEM] 20160729 No Record
* [SYSTEM] 20160730 No Recorda
* [SYSTEM] 20160731 No Record
* [SYSTEM] 20160803 No Record
* [SYSTEM] 20160804 No Record
* [SYSTEM] 20160806 No Record
* [SYSTEM] 20160807 No Record
* [SYSTEM] 20160808 No Record
* [SYSTEM] 20160809 No Record
* [SYSTEM] 20160810 No Record
* [SYSTEM] 20160811 No Record
* [SYSTEM] 20160927 Incomplete Data
* [SYSTEM] 20160928 Incomplete Data
* [SYSTEM] 20161007 No Record
* [SYSTEM] 20161008 No Record
* [SYSTEM] 20161010 No Record
* [SYSTEM] 20161011 No Record


## 5月至10月共有 184 天
### 記錄到問題資料有 28 筆
* incomplet data: 2
* no record: 26
### 記錄到正常資料有 156 筆

# 2015 小結報[PASS]
* [SYSTEM] 20150503 No Record
* [SYSTEM] 20150510 No Record
* [SYSTEM] 20150511 No Record
* [SYSTEM] 20150727 No Record
* [SYSTEM] 20150731 Incomplete Data
* [SYSTEM] 20150810 No Record
* [SYSTEM] 20150811 No Record
* [SYSTEM] 20150818 Incomplete Data
* [SYSTEM] 20150824 Incomplete Data
* [SYSTEM] 20150829 Incomplete Data
* [SYSTEM] 20150929 Incomplete Data
* [SYSTEM] 20151002 No Record
* [SYSTEM] 20151023 Incomplete Data


## 5月至10月共有 184 天
### 記錄到問題資料有 13 筆
* incomplet data: 6
* no record: 7
### 記錄到正常資料有 171 筆